BaikalDataGenerator/Source/render.cpp

```diff
@@ -0,0 +1,437 @@
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+********************************************************************/
+
+#include "render.h"
+#include "scene_io.h"
+#include "material_io.h"
+#include "SceneGraph/light.h"
+#include "Output/clwoutput.h"
+#include "BaikalIO/image_io.h"
+
+#include <fstream>
+#include "XML/tinyxml2.h"
+
+using namespace Baikal;
+
+static std::uint32_t constexpr kNumIterations = 4096;
+
+struct OutputDesc
+{
+    Renderer::OutputType type;
+    std::string name;
+    // extansion of the file to save
+    std::string file_ext;
+    int bpp;
+};
+
+// if you need to add new output for saving to disk
+// just put its description in thic collection
+std::vector<OutputDesc> outputs_collection = { { Renderer::OutputType::kColor, "color", "png", 16 },
+                                               { Renderer::OutputType::kViewShadingNormal, "view_shading_normal", "png", 8 },
+                                               { Renderer::OutputType::kDepth, "depth", "png", 16 },
+                                               { Renderer::OutputType::kAlbedo, "albedo", "png", 8 },
+                                               { Renderer::OutputType::kGloss, "gloss", "png", 8 } };
+
+static bool operator != (RadeonRays::float3 left, RadeonRays::float3 right)
+{
+    return (left.x != right.x) ||
+        (left.y != right.y) ||
+        (left.z != right.z);
+}
+
+Render::Render(const std::string &file_name,
+               const std::string &path,
+               std::uint32_t output_width,
+               std::uint32_t output_height)
+{
+    std::vector<CLWPlatform> platforms;
+    CLWPlatform::CreateAllPlatforms(platforms);
+
+    int platform_index = 0;
+
+    for (auto j = 0u; j < platforms.size(); ++j)
+    {
+        for (auto i = 0u; i < platforms[j].GetDeviceCount(); ++i)
+        {
+            if (platforms[j].GetDevice(i).GetType() == CL_DEVICE_TYPE_GPU)
+            {
+                platform_index = j;
+                break;
+            }
+        }
+    }
+
+    int device_index = 0;
+
+    for (auto i = 0u; i < platforms[platform_index].GetDeviceCount(); ++i)
+    {
+        if (platforms[platform_index].GetDevice(i).GetType() == CL_DEVICE_TYPE_GPU)
+        {
+            device_index = i;
+            break;
+        }
+    }
```

Comment:
How about std::find_if instead of such long cycles ?

Reply:
Done